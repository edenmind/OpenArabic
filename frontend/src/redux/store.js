import {
  arabicSentenceReducer,
  arabicTextReducer,
  arabicWordsReducer,
  authorReducer,
  categoryReducer,
  englishSentenceReducer,
  englishTextReducer,
  englishWordsReducer,
  sourceReducer,
  textReducer,
  wordByWordReducer,
} from './reducers'

import { configureStore } from '@reduxjs/toolkit'

export default configureStore({
  reducer: {
    title: textReducer,
    category: categoryReducer,
    author: authorReducer,
    source: sourceReducer,
    englishText: englishTextReducer,
    arabicText: arabicTextReducer,
    arabicWords: arabicWordsReducer,
    englishWords: englishWordsReducer,
    arabicSentence: arabicSentenceReducer,
    englishSentence: englishSentenceReducer,
    wordByWord: wordByWordReducer,
  },
})
