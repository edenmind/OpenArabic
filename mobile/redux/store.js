import { configureStore, combineReducers } from '@reduxjs/toolkit'

import { categoriesReducer, textReducer, textsReducer, UIStateReducer, wordsReducer } from './reducers.js'

const rootReducer = combineReducers({
  categories: categoriesReducer,
  text: textReducer,
  texts: textsReducer,
  textLoading: UIStateReducer,
  textsLoading: UIStateReducer,
  arabicFontSize: UIStateReducer,
  englishFontSize: UIStateReducer,
  isTransliterationOn: UIStateReducer,
  arabicFontName: UIStateReducer,
  practicingWords: UIStateReducer,
  words: wordsReducer,
  isDarkMode: UIStateReducer
})

const store = configureStore({
  reducer: rootReducer
})

export { store }
