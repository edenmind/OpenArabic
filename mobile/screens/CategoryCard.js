/* eslint-disable import/namespace */
import * as utility from '../services/UtilityService'

import { Caption, Card, Paragraph } from 'react-native-paper'

import { COLORS } from '../constants/colors'
import PressableOpacity from '../components/PressableOpacity'
import PropTypes from 'prop-types'
import React from 'react'
import { SCREENS } from '../constants/screens'
import { StyleSheet } from 'react-native'

const style = StyleSheet.create({
  arabic: {
    direction: 'rtl',
    fontSize: 18,
    lineHeight: 30,
    writingDirection: 'rtl'
  },
  card: {
    backgroundColor: COLORS.shinyOlive,
    marginBottom: 10,
    marginLeft: 10,
    marginRight: 10,
    marginTop: 10
  },
  cardAction: {
    marginLeft: 10,
    paddingBottom: 20,
    paddingTop: 20
  }
})

const prepareIngress = (text, length) => {
  const noLineBreaks = utility.removeLineBreak(text)
  return utility.truncate(noLineBreaks, length)
}

export default function CategoryCard(props) {
  const category = `#${props.text.category.toLowerCase()}`

  return (
    <PressableOpacity
      onPress={() => {
        props.setShouldReload(false)
        props.navigation.navigate(SCREENS.textScreen, {
          id: props.text.id
        })
      }}>
      <Card style={style.card} testID="textCard">
        <Card.Cover source={{ uri: 'https://picsum.photos/300' }} />
        <Card.Title
          title={props.text.title}
          subtitle={`${props.text.author} â€” ${props.text.source}`}
        />
        <Card.Content>
          <Paragraph>{prepareIngress(props.text.texts.english, 125)}</Paragraph>
          <Paragraph style={style.arabic}>{prepareIngress(props.text.texts.arabic, 100)}</Paragraph>
        </Card.Content>
        <Card.Actions style={style.cardAction}>
          <Caption>{category}</Caption>
        </Card.Actions>
      </Card>
    </PressableOpacity>
  )
}

CategoryCard.propTypes = {
  setShouldReload: PropTypes.func.isRequired,
  navigation: PropTypes.object,
  text: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
    author: PropTypes.string,
    source: PropTypes.string,
    texts: PropTypes.object,
    category: PropTypes.string
  })
}
