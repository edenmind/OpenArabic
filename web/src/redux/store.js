import * as reducers from './reducers'

import { configureStore } from '@reduxjs/toolkit'

export default configureStore({
  reducer: {
    text: reducers.textReducer,
  },
})
