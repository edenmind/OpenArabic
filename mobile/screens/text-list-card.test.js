import { describe, expect, jest } from '@jest/globals'
import CategoryCard from './text-list-card.js'
import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow' // ES6
import { Provider } from 'react-redux'
import { store } from '../redux/store.js'
import { NavigationContainer } from '@react-navigation/native'

jest.useFakeTimers()

describe('<CategoryCard />', () => {
  const textMock = {
    title: 'someTitle',
    id: 'someId',
    category: 'someCategory',
    author: 'someAuthor',
    texts: {
      english: 'someEnglish',
      arabic: 'someArabic'
    }
  }

  test('renders correctly', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={textMock} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no image', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, image: '' }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no author', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, author: '' }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no source', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, source: '' }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no category', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, category: '' }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no texts', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, texts: {} }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no english text', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, texts: { arabic: 'someArabic' } }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no arabic text', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, texts: { english: 'someEnglish' } }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  test('renders correctly with no texts', () => {
    const renderer = new ShallowRenderer()
    const tree = renderer.render(
      <Provider store={store}>
        <NavigationContainer>
          <CategoryCard text={{ ...textMock, texts: {} }} setShouldReload={() => {}} />
        </NavigationContainer>
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })
})
