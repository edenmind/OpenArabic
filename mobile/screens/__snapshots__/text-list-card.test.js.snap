// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryCard /> renders correctly 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "texts": {
            "arabic": "someArabic",
            "english": "someEnglish",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no arabic text 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "texts": {
            "english": "someEnglish",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no author 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "",
          "category": "someCategory",
          "id": "someId",
          "texts": {
            "arabic": "someArabic",
            "english": "someEnglish",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no category 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "",
          "id": "someId",
          "texts": {
            "arabic": "someArabic",
            "english": "someEnglish",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no english text 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "texts": {
            "arabic": "someArabic",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no image 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "image": "",
          "texts": {
            "arabic": "someArabic",
            "english": "someEnglish",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no source 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "source": "",
          "texts": {
            "arabic": "someArabic",
            "english": "someEnglish",
          },
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no texts 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "texts": {},
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;

exports[`<CategoryCard /> renders correctly with no texts 2`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "store": {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <TextListCard
      setShouldReload={[Function]}
      text={
        {
          "author": "someAuthor",
          "category": "someCategory",
          "id": "someId",
          "texts": {},
          "title": "someTitle",
        }
      }
    />
  </ForwardRef(NavigationContainerInner)>
</Context.Provider>
`;
